File: app/src/main/java/com/sandra/tasky/adapter/HomeListAdapter.java
package com.sandra.tasky.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.TextView;

import com.sandra.tasky.R;
import com.sandra.tasky.TaskyConstants;
import com.sandra.tasky.db.TaskDatabase;
import com.sandra.tasky.entity.SimpleTask;
import com.sandra.tasky.utils.TimeUtils;

import java.util.List;

public class HomeListAdapter extends BaseAdapter {
    private Context context;
    private List<SimpleTask> taskList;

    public HomeListAdapter(Context context, List<SimpleTask> taskList) {
        this.context = context;
        this.taskList = taskList;
    }

    @Override
    public int getCount() {
        return taskList.size();
    }

    @Override
    public Object getItem(int position) {
        return taskList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, final View convertView, ViewGroup parent) {
        View listView;
        final SimpleTask task = taskList.get(position);

        listView = LayoutInflater.from(context).inflate(R.layout.item_home_screen, parent, false);

        TextView title = listView.findViewById(R.id.tw_title);
        title.setText(cutText(task.getTitle(), TaskyConstants.MAX_TITLE_LENGTH));

        final CheckBox checkBox = listView.findViewById(R.id.check_box);
        checkBox.setChecked(task.isCompleted());
        checkBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                task.setCompleted(checkBox.isChecked());
                TaskDatabase db = new TaskDatabase(context);
                db.updateTask(task);
                notifyDataSetChanged();
            }
        });

        TextView note = listView.findViewById(R.id.tw_note);
        if (!task.getNote().isEmpty()) {
            note.setText(cutText(task.getNote(), TaskyConstants.MAX_TEXT_LENGTH));
        } else {
            note.setVisibility(View.GONE);
        }

        TextView dueDate = listView.findViewById(R.id.tw_due_date);
        if (task.getDueDate() != null) {
            if (!TimeUtils.isInFuture(task)) {
                dueDate.setText(R.string.expired);
            } else {
                dueDate.setText(task.isTimePresent() ? task.parseDateTime() : task.parseDate());
            }
        } else {
            dueDate.setVisibility(View.GONE);
        }

        return listView;
    }

    private String cutText(String text, int limit) {
        return text.length() > limit ? text.substring(0, limit) + "..." : text;
    }
}

File: app/src/main/java/com/sandra/tasky/activities/home/HomeListAdapter.kt
package com.sandra.tasky.activities.home

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.sandra.tasky.R
import com.sandra.tasky.db.DatabaseWrapper
import com.sandra.tasky.entity.SimpleTask
import com.sandra.tasky.utils.capitalFirstLetter
import com.sandra.tasky.utils.hide
import kotlinx.android.synthetic.main.item_task.view.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class HomeListAdapter(private val context: Context, private val taskList: List<SimpleTask>) : BaseAdapter() {

    override fun getCount(): Int {
        return taskList.size
    }

    override fun getItem(position: Int): Any {
        return taskList[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val itemView = convertView?: LayoutInflater.from(context).inflate(R.layout.item_task, parent, false)

        val task = taskList[position]

        itemView.tvTitle.text = cutText(task.title, MAX_TITLE_LENGTH)
        val checkBox = itemView.cbCompleted
        checkBox.isChecked = task.isCompleted
        checkBox.setOnClickListener {
            task.isCompleted = checkBox.isChecked

            CoroutineScope(Dispatchers.Main).launch {
                DatabaseWrapper.updateTask(context, task)
                notifyDataSetChanged()
            }
        }
        if (task.note.isNotEmpty()) {
            itemView.tvNote.text = cutText(task.note, MAX_TEXT_LENGTH)
        } else {
            itemView.tvNote.hide()
        }
        if (task.dueDate != null) {
            if (task.isExpired) {
                itemView.tvDueDate.text = context.getString(R.string.expired).capitalFirstLetter()
            } else {
                itemView.tvDueDate.text = if (task.isTimePresent) task.parseDateTime() else task.parseDate()
            }
        } else {
            itemView.tvDueDate.hide()
        }
        return itemView
    }

    private fun cutText(text: String?, limit: Int): String {
        return (if (text!!.length > limit) text.substring(0, limit) + "..." else text)
    }

    companion object {
        private const val MAX_TITLE_LENGTH = 70
        private const val MAX_TEXT_LENGTH = 100
    }

}